Sauna refactor

Esinko opened the pull request:
Draft pull request for the sauna refactor by Elias (awwry).

hunaja committed:
chore: introduce relevant eslint overrides

hunaja committed:
refactor(sauna!): db structure, models, routes

hunaja committed:
refactor: introduce dotenv and connectDatabase.js

hunaja committed:
feat(sauna): throw yup out and just use mongo

supertassu commented on sauna/src/util/connectDatabase.js:
```diff
@@ -0,0 +1,20 @@
+import mongoose from 'mongoose';
+import dotenv from 'dotenv';
+
+dotenv.config();
```
this should not be in "connectDatabase", I'd put it in main.js for now

supertassu commented on sauna/src/controllers/heartbeat.js:
```diff
@@ -0,0 +1,59 @@
+            projectMatch.hours.push(date.getHours());
+        }
+
+        projectMatch.totalTime += 2;
```
should be configurable in the request body

supertassu commented on sauna/src/controllers/heartbeat.js:
```diff
@@ -0,0 +1,59 @@
+    const depopulated = user.depopulate('editedProjects');
+
+    if (projectMatch) {
+        if (!projectMatch.editors.includes(editor)) projectMatch.editors.push(editor);
+
+        if (!projectMatch.hours.includes(date.getUTCHours())) {
+            projectMatch.hours.push(date.getHours());
+        }
```
please be consistent with wrapping statements with `{}` or not

supertassu commented on sauna/src/controllers/users.js:
```diff
@@ -0,0 +1,78 @@
+    const users = await User.find({})
+        .populate('editedProjects');
+
+    const sortedUsers = users.sort((a, b) => countTotalTime(b) - countTotalTime(a)).splice(0, 10);
```
not familiar with Mongo, but could this be done on database side with some sort of index? this will get expensive quickly. If not it should be cached for some period of time

supertassu commented on sauna/src/controllers/users.js:
```diff
@@ -0,0 +1,78 @@
+    const { id } = req.params;
+
+    const user = await User.findById(id).populate('editedProjects');
+    if (!user) return res.status(403);
```
404

supertassu commented on sauna/src/controllers/users.js:
```diff
@@ -0,0 +1,78 @@
+
+    let totalTime = 0;
+
+    const projects = user.editedProjects.reduce((a, e) => {
```
please use more descriptive variable names

supertassu commented on sauna/src/controllers/users.js:
```diff
@@ -0,0 +1,78 @@
+/** Require given user to exist */
+router.use('/:id', async (req, res, next) => {
+    if (!mongoose.Types.ObjectId.isValid(req.params.id)) {
+        return res.sendStatus(400);
```
please add informative error message in addition to status code

supertassu commented on sauna/src/controllers/heartbeat.js:
```diff
@@ -0,0 +1,59 @@
+        });
+
+        const saveSuccess = await editedProject.save();
+        if (!saveSuccess) return res.status(400);
```
500, unless known to be an user error

supertassu commented on sauna/src/controllers/heartbeat.js:
```diff
@@ -0,0 +1,59 @@
+        projectMatch.totalTime += 2;
+
+        const saveSuccess = await projectMatch.save();
+        if (!saveSuccess) return res.status(400);
```
500, unless known to be an user error

supertassu reviewed (changes requested):
few comments based on a surface look, still untested

hunaja committed:
refractor(sauna): heartbeat router style and db

Chicken commented on sauna/package-lock.json:
```diff
@@ -1,2935 +1,8 @@
   "name": "sauna",
   "version": "0.0.1",
-  "lockfileVersion": 2,
+  "lockfileVersion": 1,
```
Update your npm installation with `npm install -g npm@latest` and reinstall packages.

ahnl commented on sauna/src/index.js:
```diff
@@ -3,48 +3,42 @@
+import usersController from './controllers/users.js';
+import heartbeatController from './controllers/heartbeat.js';
+
+const port = 8080;
```
Because sauna is also serving React content, it would overall make more sense to use the port 80.

Chicken replied:
It would make even more sense to add the port to environmental variables or config.

Chicken commented on sauna/src/controllers/heartbeat.js:
```diff
@@ -0,0 +1,59 @@
+    if (!userWithEdits) return res.status(403);
+
+    const { test = false, project = null, editor = null } = req.body;
+    if (!project || !editor) return res.status(400);
```
Check the received variable types too.

Chicken reviewed (changes requested):
Probably better than what I would be able to do lol

ahnl commented on sauna/src/index.js:
```diff
@@ -3,48 +3,42 @@
+app.use('/api/heartbeat', heartbeatController);
 
+// React content
+app.use((_, res) => res.sendFile('index.html', {root: '../salmiakki/build/'}));
 app.use('/', express.static('../salmiakki/build/'));
```
Not working, they're just wrong way around.

```js
app.use('/', express.static('../salmiakki/build/'));
app.use((_, res) => res.sendFile('index.html', {root: '../salmiakki/build/'}));
```

ahnl committed:
fix(sauna): React serving and HTTP port

Esinko committed:
fix: Exit code & API error message

users.js: Added error message.
index.js: Error code changed to -1 when exiting the process due to an error.

Esinko committed:
fix: General fixes & Error handling for index.js

- Some general fixes to style and variable names.
- Imported mongoose.
- Handled "reconnectFailed" event for mongoose.

Esinko committed:
chore: Use destructuring for mongoose

Esinko committed:
chore: Circumvent top-level await

hunaja committed:
fix: add validators to some db schema fields

hunaja committed:
chore: newer package-lock.json

hunaja committed:
fix: configurable editing time

hunaja committed:
fix: proper error handling

Chicken commented on sauna/src/models/user.js:
```diff
@@ -0,0 +1,23 @@
+    apiKey: {
+        type: String,
+        required: true,
+    },
```
Chances of duplicate api keys is small but it doesn't hurt to enforce it.
```suggestion
    apiKey: {
        type: String,
        required: true,
        unique: true,
    },
```

supertassu commented on sauna/src/index.js:
```diff
@@ -3,48 +3,61 @@
 
-app.listen(80, () => {
-    console.log('Webserver up and running on port 80');
+app.listen(port, () => {
+    console.log(`Webserver up and running on port ${port}.`);
 });
```
This should be only after it's connected to the database (in the async function above)

Chicken commented on sauna/src/controllers/users.js:
```diff
@@ -0,0 +1,89 @@
+        return editAccumulator;
+    }, {});
+
+    return res.json({
+        discordId: user.discordId,
+        apiKey: user.apiKey,
+        totalTime,
+        projects,
+    });
```
Why is the api key returned?!?!
```suggestion
    return res.json({
        discordId: user.discordId,
        totalTime,
        projects,
    });
```

ahnl committed:
Update sauna/src/models/user.js

Co-authored-by: Antti <antti@antti.codes>

ahnl committed:
enhancement(sauna): Run express server after database connection

ahnl committed:
enhancement(sauna): Remove api key from respone

Co-authored-by: Antti <antti@antti.codes>

Chicken committed:
fix(sauna): dotenv needs to be used before accesing variables

supertassu committed:
fix(sauna): do not serve the frontend
